client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/classes/URLClassReplacement.java

```diff
@@ -37,8 +38,38 @@ public static URLConnection openConnection(URL caller) throws java.io.IOExceptio
           Add the external service hostname to the ExecutionTracer
           */
         if (caller.getProtocol().equals("http") || caller.getProtocol().equals("https")) {
-            ExternalServiceInfo remoteHostInfo = new ExternalServiceInfo(caller.getProtocol(), caller.getHost(), caller.getPort());
+            int port = caller.getPort();
+            String protocol = caller.getProtocol();
+
+            // Unless the port number is specified, the default will be -1.
+            // Which indicates that the port should be assigned according to the
+            // protocol. Since the URLConnection openConnection is an abstract, this
+            // assignment will be handled under the respective implementation.
+            // Here it's manually handled assuming these default will never change. :)
+            if (port == -1) {
+                switch (protocol) {
+                    case "https":
+                        port = 443;
+                        break;
+                    case "http":
+                        port = 80;
+                        break;
+                }
+            }
+            ExternalServiceInfo remoteHostInfo = new ExternalServiceInfo(protocol, caller.getHost(), port);
             ExecutionTracer.addExternalServiceHost(remoteHostInfo);
+
+            if (ExecutionTracer.hasExternalMapping(caller.getHost())) {
+                String ip  = ExecutionTracer.getExternalMapping(caller.getHost());
+
+                // Usage of ports below 1024 require root privileges to run
+                String url = "http://" + ip + ":" + caller.getPort() + caller.getPath();
+
+                URL newURL = new URL(url);
+                return newURL.openConnection();
+            } else {
+                throw new UnknownHostException("There is no WireMock initiated for this hostname");
```

Comment:
can leave a TODO here, stating the message will likely need to be changed in the future.
if host is not reachable, exeption is thrown, and could lead to 500 with the stack-trace in response. in such case, assertions would be generated on response, which will then fail when running tests as the messages would be different when running without instrumentation. we ll likely see this problem when we will run the generated tests, and think of a solution then